src/bidiMapper/modules/context/NavigationTracker.spec.ts

```diff
@@ -0,0 +1,394 @@
+/*
+ *  Copyright 2024 Google LLC.
+ *  Copyright (c) Microsoft Corporation.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+import {assert} from 'chai';
+import sinon, {type SinonStubbedInstance} from 'sinon';
+
+import {ChromiumBidi} from '../../../protocol/protocol.js';
+import {EventManager} from '../session/EventManager.js';
+
+import {NavigationEventName, NavigationTracker} from './NavigationTracker.js';
+
+// keep-sorted start block=yes
+const ANOTHER_LOADER_ID = 'ANOTHER_LOADER_ID';
+const ANOTHER_URL = 'ANOTHER_URL';
+const BROWSING_CONTEXT_ID = 'browsingContextId';
+const ERROR_MESSAGE = 'ERROR_MESSAGE';
+const INITIAL_URL = 'INITIAL_URL';
+const LOADER_ID = 'LOADER_ID';
+const SOME_URL = 'SOME_URL';
+const YET_ANOTHER_URL = 'YET_ANOTHER_URL';
+// keep-sorted end
+
+describe('NavigationTracker', () => {
+  let navigationTracker: NavigationTracker;
+  let eventManager: SinonStubbedInstance<EventManager>;
+  let initialNavigationId: string;
+
+  async function assertNoNavigationEvents() {
+    await new Promise((resolve) => setTimeout(resolve, 1));
```

Comment:
I think draining of the microtask queue masks the issues of when events have to happen

Reply:
I would suggest we remove this and make the assertion sync